File: publisher.py
import zmq
import cv2
import numpy as np
import time
import json  # Add import for JSON handling

class Publisher:
    def __init__(self, address="tcp://*:5555", topic="camera/image_raw", chunk_size=100000):
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.PUB)
        self.socket.bind(address)
        
        self.topic = topic.encode()
        self.chunk_size = chunk_size  # Tamaño de fragmento en bytes

    def publish_image(self, image_path=None, use_camera=False):
        """ Captura una imagen de un archivo o una cámara y la publica embebida en un mensaje binario. """
        if use_camera:
            cap = cv2.VideoCapture(0)
            ret, frame = cap.read()
            cap.release()
            if not ret:
                print("❌ Error capturando imagen de la cámara.")
                return
        else:
            frame = cv2.imread(image_path)
            if frame is None:
                print(f"❌ No se pudo cargar la imagen: {image_path}")
                return

        _, buffer = cv2.imencode(".jpg", frame)
        image_bytes = buffer.tobytes()

        message = {
            "type": "image",
            "format": "jpg",
            "width": frame.shape[1],
            "height": frame.shape[0]
        }
        message_bytes = json.dumps(message).encode('utf-8') + b'\x00' + image_bytes

        self.publish_message(message_bytes)

    def publish_message(self, message_bytes):
        """ Publica un mensaje binario fragmentado. """
        num_chunks = len(message_bytes) // self.chunk_size + 1

        self.total_bytes_sent = 0  # Initialize total bytes sent

        for i in range(num_chunks):
            chunk = message_bytes[i * self.chunk_size: (i + 1) * self.chunk_size]
            self.socket.send_multipart([self.topic, str(i).encode(), str(num_chunks).encode(), chunk])
            self.total_bytes_sent += len(chunk)

        print(f"📤 Mensaje binario publicado en {num_chunks} fragmentos.")

    def close(self):
        """ Cierra la conexión ZeroMQ. """
        self.socket.close()
        self.context.term()

File: src/go/publisher/publisher.go
package publisher

import (
    "encoding/json"
    "fmt"
    "github.com/pebbe/zmq4"
    "github.com/disintegration/imaging"
    "image"
    "image/jpeg"
    "os"
    "time"
)

type Publisher struct {
    socket *zmq4.Socket
    topic  string
}

type ImageMetadata struct {
    Format string `json:"format"`
    Width  int    `json:"width"`
    Height int    `json:"height"`
    Size   int    `json:"size"`
}

type Message struct {
    Type   string          `json:"type"`
    Count  int             `json:"count"`
    Images []ImageMetadata `json:"images"`
    Data   map[string]interface{} `json:"data"`
}

func NewPublisher(address, topic string) (*Publisher, error) {
    context, err := zmq4.NewContext()
    if err != nil {
        return nil, err
    }
    socket, err := context.NewSocket(zmq4.PUB)
    if err != nil {
        return nil, err
    }
    err = socket.Bind(address)
    if err != nil {
        return nil, err
    }
    return &Publisher{socket: socket, topic: topic}, nil
}

func (p *Publisher) BuildMessage(frames []image.Image, data map[string]interface{}) ([]byte, error) {
    var imagesMetadata []ImageMetadata
    var imagesData []byte

    for _, frame := range frames {
        buffer := new(bytes.Buffer)
        err := jpeg.Encode(buffer, frame, nil)
        if err != nil {
            return nil, err
        }
        imageBytes := buffer.Bytes()
        imageMetadata := ImageMetadata{
            Format: "jpg",
            Width:  frame.Bounds().Dx(),
            Height: frame.Bounds().Dy(),
            Size:   len(imageBytes),
        }
        imagesMetadata = append(imagesMetadata, imageMetadata)
        imagesData = append(imagesData, imageBytes...)
    }

    message := Message{
        Type:   "images",
        Count:  len(frames),
        Images: imagesMetadata,
        Data:   data,
    }
    messageJSON, err := json.Marshal(message)
    if err != nil {
        return nil, err
    }
    return append(messageJSON, append([]byte{0}, imagesData...)...), nil
}

func (p *Publisher) PublishMessage(messageBytes []byte) error {
    _, err := p.socket.SendMessage(p.topic, messageBytes)
    return err
}

func (p *Publisher) Close() {
    p.socket.Close()
}
